module/evdi_modeset.c

```diff
@@ -89,89 +89,6 @@ static void evdi_crtc_atomic_flush(
 	crtc_state->event = NULL;
 }
 
-static void evdi_mark_full_screen_dirty(struct evdi_device *evdi)
```

Comment:
Please do not remove old implementation that is using evdi_crtc_cursor_set, evdi_crtc_cursor_move callbacks, as those are used by XServer and older mutter. You can switch to other compositor at runtime. So we need both implementations.

What DRM bugcheck is failing?

Reply:
We need to understand preferable way of switching cursor implementation. Can you check how it is done in modules from latest kernel (e.g udl, exynos, vgm)?
Does they provide just one implementation?

Reply:
When both mechanisms are enabled DRM triggers assertion in drm_mode_config.c:651, see attached:
[drm_assertion.txt](https://github.com/DisplayLink/evdi/files/7289166/drm_assertion.txt)

I looked at some drivers (kernel 5.14 debian/bookworm). UDL does not support hardware cursor at all. Most of the drivers use cursor plane by now (i915, amdgpu, virtio, vmwgfx...), few older ones still seem to work with legacy API (radeon, gma500). 

It seems only Qualcomm's mdp5 driver enables both mechanisms in kernel 5.14 (msm/disp/mdp5/mdp5_crtc.c:1315) The very same issue I am reporting here has been raised 10 days ago! (https://www.mail-archive.com/dri-devel@lists.freedesktop.org/msg367990.html). The fix will be to turn on cursor_plane if it is supported by HW and not legacy API.

Are you sure that X11 and other configurations that you need to support will not work with cursor planes? 
Maybe that needs testing? 

Anyway it seems that if the choice has to be made it will have to be made during probe and it will not be able to reconfigure during runtime if the user switches compositor. 

I would think a kernel version check to choose between the two will suffice? Which kernel version would that be?
I am open to suggestions.
Thanks.

Reply:
Ok. TBH I can't tell exactly which versions of XServer was using legacy callbacks. But I would prefer to keep old implementation for older systems for a while until we remove it completely.

Generally I prefer to avoid any dynamic switches if possible. It would be reasonable to use cursor planes from kernel version which introduced that bugcheck. 

But we need to be sure that it is fine with xserver.
Can you test that with fallowing scenarios:
 - XServer on Ubuntu 21.04 + evdi with cursor planes
 - XServer on Ubuntu 20.04 + evdi with cursor planes
 
Many Thanks Piotr!

Reply:
The DRM check has been added in kernel 5.10, so I updated the patch to enable cursor planes from that version.
I tested Ubuntu 20.04 and 21.04. Both use kernel 5.11 so they switch to the new mechanism. 
Both compositors (X11 and mutter/wayland) detect cursor planes and start using them. So the proposed change works.

I noticed that with X11 the user-space driver is constantly reading pixels from EVDI (painter_grabpix), while the system is displaying static wallpaper only. This is happening on both Ubuntu versions and irrespective of which cursor mechanism is used. Wayland does not have that problem. So I reckon something is broken with X11 path. It has nothing to do with proposed change though.

Thanks.